src/file/BaseTagFS.cpp

```diff
@@ -0,0 +1,205 @@
+}
+
+
+void BaseTagFS::addReference(const std::string &name_or_id) {
+    if (name_or_id.empty())
+        throw EmptyString("addReference");
+
+    if (!block()->hasDataArray(name_or_id))
+        throw std::runtime_error("BaseTagFS::addReference: DataArray not found in block!");
+
+    auto target = dynamic_pointer_cast<DataArrayFS>(block()->getDataArray(name_or_id));
+    refs_group.createDirectoryLink(target->location(), target->id());
+}
+
+
+bool BaseTagFS::removeReference(const std::string &name_or_id) {
+    return refs_group.removeObjectByNameOrAttribute("name", name_or_id);
+}
+
+
+void BaseTagFS::references(const std::vector<DataArray> &refs_new) {
+    // extract vectors of names from vectors of new & old references
+    std::vector<std::string> names_new(refs_new.size());
+    transform(refs_new.begin(), refs_new.end(), names_new.begin(), util::toName<DataArray>);
+    //FIXME: issue 473
+    std::vector<DataArray> refs_old(static_cast<size_t>(referenceCount()));
+    for (size_t i = 0; i < refs_old.size(); i++) refs_old[i] = getReference(i);
+    std::vector<std::string> names_old(refs_old.size());
+    transform(refs_old.begin(), refs_old.end(), names_old.begin(), util::toName<DataArray>);
+
+    // sort them
+    std::sort(names_new.begin(), names_new.end());
+    std::sort(names_new.begin(), names_new.end());
+
+    // get names only in names_new (add), names only in names_old (remove) & ignore rest
+    std::vector<std::string> names_add;
+    std::vector<std::string> names_rem;
+    std::set_difference(names_new.begin(), names_new.end(), names_old.begin(), names_old.end(),
+                        std::inserter(names_add, names_add.begin()));
+    std::set_difference(names_old.begin(), names_old.end(), names_new.begin(), names_new.end(),
+                        std::inserter(names_rem, names_rem.begin()));
+
+    // check if all new references exist & add sources
+    auto blck = dynamic_pointer_cast<BlockFS>(block());
+    for (auto name : names_add) {
+        if (!blck->hasDataArray(name))
+            throw std::runtime_error("One or more data arrays do not exist in this block!");
+        addReference(blck->getDataArray(name)->id());
+    }
+    // remove references
+    for (auto name : names_rem) {
+        if (!blck->hasDataArray(name))
+            removeReference(blck->getDataArray(name)->id());
+    }
+}
+
+//--------------------------------------------------
+// Methods concerning features.
+//--------------------------------------------------
+
+bool BaseTagFS::hasFeature(const string &name_or_id) const {
+    return !feature_group.findByNameOrAttribute("entity_id", name_or_id)->empty();
+}
+
+
+ndsize_t BaseTagFS::featureCount() const {
+    return feature_group.subdirCount();
+}
+
+
+shared_ptr<IFeature> BaseTagFS::getFeature(const std::string &name_or_id) const {
+    shared_ptr<IFeature> feature;
+    boost::optional<boost::filesystem::path> p = feature_group.findByNameOrAttribute("name", name_or_id);
```

Comment:
put `namespace bfs = boost::filesystem` somewhere above and then `bfs::path` maybe?